internal/ext/slicesx/queue.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2020-2025 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package slicesx
+
+import (
+	"fmt"
+	"iter"
+	"slices"
+
+	"github.com/bufbuild/protocompile/internal/ext/bitsx"
+	"github.com/bufbuild/protocompile/internal/ext/iterx"
+)
+
+// Set to true to enable debug printing for queues.
+const debugQueue = false
+
+// Queue is a ring buffer.
+//
+// Values can be pushed and popped from either the front or the back of the
+// buffer, making it usable as a double-ended queue.
+//
+// A zero [Queue] is empty and ready to use.
+type Queue[E any] struct {
+	buf        []E // Invariant: cap(buf) is always a power of 2, or zero.
+	start, end int
+}
+
+// NewRing returns a [Queue] with the given capacity.
+func NewRing[E any](capacity int) *Queue[E] {
```

Comment:
NewRing -> NewQueue (josh made me rename the type to Queue, it was originally Ring)